### ISSUE TITLE
AttributeError: 'Conv2D' object has no attribute 'shape' and ValueError: You are trying to load a weight file containing 1 layers into a model with 19 layers.

### ISSUE THREAD

**Issue by Space-Fighter** [tensorflow/tensorflow#52184]
TENSORFLOW VERSION: 2.1.0
Hey there i am trying to load a model like the following - 
Below is my model architecture - 
```py
def down_block(x, filters, kernel_size=(3, 3), padding="same", strides=1, input = False):
    if input == False:
        c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(x)
    else:
        c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu", input_shape=(HEIGHT, WIDTH, 3))
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(c)
    p = MaxPool2D((2, 2), (2, 2))(c)
    return c, p

def up_block(x, skip, filters, kernel_size=(3, 3), padding="same", strides=1):
    us = UpSampling2D((2, 2))(x)
    concat = Concatenate()([us, skip])
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(concat)
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(c)
    return c

def bottleneck(x, filters, kernel_size=(3, 3), padding="same", strides=1):
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(x)
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(c)
    return c

def UNet():
    f = [16, 32, 64, 128, 256]
 
    

    c1, p1 = down_block(None, f[0], input = True) #128 -> 64
    c2, p2 = down_block(p1, f[1]) #64 -> 32
    c3, p3 = down_block(p2, f[2]) #32 -> 16
    c4, p4 = down_block(p3, f[3]) #16->8
    
    bn = bottleneck(p4, f[4])
    
    u1 = up_block(bn, c4, f[3]) #8 -> 16
    u2 = up_block(u1, c3, f[2]) #16 -> 32
    u3 = up_block(u2, c2, f[1]) #32 -> 64
    u4 = up_block(u3, c1, f[0]) #64 -> 128
    
    outputs = Conv2D(13, (1, 1), padding="same", activation="sigmoid")(u4)
    model = Model(inputs, outputs)
    return model
```
Then i make a variable named unet and pass in the model architecture - 
```py
unet = UNet()
unet.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["acc"])
unet.summary()
```

I get the following error - 
```py
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-26-33ba33cd9519> in <module>
----> 1 unet = UNet()
      2 unet.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["acc"])
      3 unet.summary()

<ipython-input-25-6aea0c9181eb> in UNet()
     25 
     26 
---> 27     c1, p1 = down_block(None, f[0], input = True) #128 -> 64
     28     c2, p2 = down_block(p1, f[1]) #64 -> 32
     29     c3, p3 = down_block(p2, f[2]) #32 -> 16

AttributeError: 'Conv2D' object has no attribute 'shape'
```

Full traceback below -
https://pastebin.com/mkTSg4bS

by the way originally i started out with the following code the thing is the actual code was -
```py
def down_block(x, filters, kernel_size=(3, 3), padding="same", strides=1):
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(x)
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(c)
    p = MaxPool2D((2, 2), (2, 2))(c)
    return c, p

def up_block(x, skip, filters, kernel_size=(3, 3), padding="same", strides=1):
    us = UpSampling2D((2, 2))(x)
    concat = Concatenate()([us, skip])
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(concat)
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(c)
    return c

def bottleneck(x, filters, kernel_size=(3, 3), padding="same", strides=1):
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(x)
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(c)
    return c

def UNet():
    f = [16, 32, 64, 128, 256]
    inputs = Input((HEIGHT,WIDTH,3))
    
    p0 = inputs
    c1, p1 = down_block(p0, f[0]) #128 -> 64
    c2, p2 = down_block(p1, f[1]) #64 -> 32
    c3, p3 = down_block(p2, f[2]) #32 -> 16
    c4, p4 = down_block(p3, f[3]) #16->8
    
    bn = bottleneck(p4, f[4])
    
    u1 = up_block(bn, c4, f[3]) #8 -> 16
    u2 = up_block(u1, c3, f[2]) #16 -> 32
    u3 = up_block(u2, c2, f[1]) #32 -> 64
    u4 = up_block(u3, c1, f[0]) #64 -> 128
    
    outputs = Conv2D(13, (1, 1), padding="same", activation="sigmoid")(u4)
    model = Model(inputs, outputs)
    return model
```
When i would try to load my model using this code- 
```py
unet = UNet()
unet.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["acc"])
unet.summary()

# unet.load_weights('unet1m.h5')
unet.load_weights('unet.h5')
```
 i would get the following  errro -
```
ValueError: You are trying to load a weight file containing 1 layers into a model with 19 layers.
```
Then i found this issue on github [ISSUE](https://github.com/keras-team/keras/issues/10417#issuecomment-415905336).
 It suggested not to use an input layer instead use input_shape in conv2d 
so i edited the code as you see in the original post
i did not want to modify my whole code so came up with a messy solution
i added an argument named input
then in the function i wrote that if input = False just do the bussiness it was doing in the original code
if  it was input = true we add that extra input_shape argument in Conv2D 
Now for not to remove the x parameter i just simply passed x as none in the input block
Then i found this issue on github. It suggested not to use an input layer instead use input_shape in conv2d 
so i edited the code as you see at the top

**FOR EXPERIMENTING WITH THE CODE **
The model -
[MODEL](https://drive.google.com/file/d/1RaRn9eI40ZDXAX-ajnCT0oWJclR8w9RF/view)
The script - 
[SCRIPT](https://gist.github.com/Space-Fighter/53d7a25dd50278c9106103dc6bb5afbd)
The training script - 
[TRAINING SCRIPT](https://github.com/srihari-humbarwadi/cityscapes-segmentation-with-Unet/blob/master/batch_training.py)
The repositry from where i have taken the model and code - 
[REPOSITRY](https://github.com/srihari-humbarwadi/cityscapes-segmentation-with-Unet)

**Comment by vulkomilev** [comment]
```
   if input == False:
        c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(x)
    else:
        c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu", input_shape=(HEIGHT, WIDTH, 3))
  
```
I think that you forgot the (x) in 
       c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu", input_shape=(HEIGHT, WIDTH, 3))

**Comment by mohantym** [comment]
Hi @Space-Fighter! 
We see that the issue template has not been filled, could you please do so as it helps us analyse the issue [tf version, steps followed before you ran into this error or stand alone code/colab gist to reproduce the issue faced].Thanks

**Comment by Space-Fighter** [comment]
> Hi @Space-Fighter! We see that the issue template has not been filled, could you please do so as it helps us analyse the issue [tf version, steps followed before you ran into this error or stand alone code/colab gist to reproduce the issue faced].Thanks!

Hey there i have added the tensorflow version rest everything is already there like steps followed before you ran into this error or stand alone code/colab gist to reproduce the issue faced].

**Comment by Space-Fighter** [comment]
@vulkomilev as i have mentioned in my post - 

originally i started out with the following code -
```py
def down_block(x, filters, kernel_size=(3, 3), padding="same", strides=1):
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(x)
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(c)
    p = MaxPool2D((2, 2), (2, 2))(c)
    return c, p

def up_block(x, skip, filters, kernel_size=(3, 3), padding="same", strides=1):
    us = UpSampling2D((2, 2))(x)
    concat = Concatenate()([us, skip])
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(concat)
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(c)
    return c

def bottleneck(x, filters, kernel_size=(3, 3), padding="same", strides=1):
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(x)
    c = Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(c)
    return c

def UNet():
    f = [16, 32, 64, 128, 256]
    inputs = Input((HEIGHT,WIDTH,3))
    
    p0 = inputs
    c1, p1 = down_block(p0, f[0]) #128 -> 64
    c2, p2 = down_block(p1, f[1]) #64 -> 32
    c3, p3 = down_block(p2, f[2]) #32 -> 16
    c4, p4 = down_block(p3, f[3]) #16->8
    
    bn = bottleneck(p4, f[4])
    
    u1 = up_block(bn, c4, f[3]) #8 -> 16
    u2 = up_block(u1, c3, f[2]) #16 -> 32
    u3 = up_block(u2, c2, f[1]) #32 -> 64
    u4 = up_block(u3, c1, f[0]) #64 -> 128
    
    outputs = Conv2D(13, (1, 1), padding="same", activation="sigmoid")(u4)
    model = Model(inputs, outputs)
    return model
```
When i would try to load my model using this code- 
```py
unet = UNet()
unet.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["acc"])
unet.summary()

# unet.load_weights('unet1m.h5')
unet.load_weights('unet.h5')
```
 i would get the following  errro -
```
ValueError: You are trying to load a weight file containing 1 layers into a model with 19 layers.
```
Then i found this issue on github. It suggested not to use an input layer instead use input_shape in conv2d 
so i edited the code as you see in the original post
i did not want to modify my whole code so came up with a messy solution
i added an argument named input
then in the function i wrote that if input = False just do the bussiness it was doing in the original code
if  it was input = true we add that extra input_shape argument in Conv2D 
Now for not to remove the x parameter i just simply passed x as none in the input block
Then i found this issue on github. It suggested not to use an input layer instead use input_shape in conv2d 
so i edited the code as you see at the top

**Comment by mohantym** [comment]
Hi @sanatmpa1! ,Could you please look at this issue , was getting different Error in TF [2.5 ](https://colab.sandbox.google.com/drive/1iXaogg_FOcSZTsckaJkKq0-UbbqwMZ6t?resourcekey=0-sNtR27kcDZSZ-KRioS-ceA#scrollTo=FsoDIPU4861i),[2.6](https://colab.research.google.com/gist/mohantym/2c0d4715bcd8038e3c0f11b7fdd230d8/github_52184.ipynb#scrollTo=BQVBCqlB8uH4) and [nightly](https://colab.research.google.com/gist/mohantym/a971b6a86fe1362c41df58039a285700/github_52184.ipynb#scrollTo=BQVBCqlB8uH4).

**Comment by sanatmpa1** [comment]
@Space-Fighter 

As per the [script](https://pastebin.com/kCr8DDLk) that you provided, You have provided input `x` as `None` in this line `c1, p1 = down_block(None, f[0], input = True) #128 -> 64` and so while trying to reproduce your code, I am getting the error `TypeError: Inputs to a layer should be tensors. Got: None` [gist here](https://colab.research.google.com/gist/sanatmpa1/e92bbc22183fc6d32e8fa8344ef0bb4e/52184.ipynb). Can you provide a github gist to reproduce the issue reported?

**Comment by Space-Fighter** [comment]
> @Space-Fighter
> 
> As per the [script](https://pastebin.com/kCr8DDLk) that you provided, You have provided input `x` as `None` in this line `c1, p1 = down_block(None, f[0], input = True) #128 -> 64` and so while trying to reproduce your code, I am getting the error `TypeError: Inputs to a layer should be tensors. Got: None` [gist here](https://colab.research.google.com/gist/sanatmpa1/e92bbc22183fc6d32e8fa8344ef0bb4e/52184.ipynb). Can you provide a github gist to reproduce the issue reported?

@sanatmpa1 oh minor correction in the post i mentioned the following line - 
![image](https://user-images.githubusercontent.com/63576756/136048511-c37f4e93-a6a9-41b0-aced-7834b057de79.png)
but i mistakenly sent my old script - 
![image](https://user-images.githubusercontent.com/63576756/136048696-6c9ebf1d-17e2-417c-afdf-54323d328824.png)


Here is the script - 
[SCRIPT](https://gist.github.com/Space-Fighter/53d7a25dd50278c9106103dc6bb5afbd)

I have updated this in the post too

**Comment by Space-Fighter** [comment]
@sanatmpa1 Trying out the first suggestion i get the following error - 
```py
AttributeError                            Traceback (most recent call last)
<ipython-input-7-33ba33cd9519> in <module>
----> 1 unet = UNet()
      2 unet.compile(optimizer="adam", loss="categorical_crossentropy", metrics=["acc"])
      3 unet.summary()

<ipython-input-6-6627c0447d90> in UNet()
     28 
     29 
---> 30     c1, p1 = down_block(None, f[0], input = True) #128 -> 64
     31     c2, p2 = down_block(p1, f[1]) #64 -> 32
     32     c3, p3 = down_block(p2, f[2]) #32 -> 16

<ipython-input-6-6627c0447d90> in down_block(x, filters, kernel_size, padding, strides, input)
      5         c = tf.keras.layers.Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(x)
      6     else:
----> 7         c = tf.keras.layers.Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu", input_shape=(512,512, 3))(x)
      8         c = tf.keras.layers.Conv2D(filters, kernel_size, padding=padding, strides=strides, activation="relu")(c)
      9         p = tf.keras.layers.MaxPool2D((2, 2), (2, 2))(c)
AttributeError: 'NoneType' object has no attribute 'shape'
```
[FULL TRACEBACK](https://gist.github.com/Space-Fighter/847755b29adb561d1fe65f0fa316db46)

Trying out the second suggestion my code becomes - 

[CODE](https://gist.github.com/Space-Fighter/7f926edf96260758f901b06ae5a72157)

Error i get - 

[ERROR](https://gist.github.com/Space-Fighter/05cbf8a19172ceab33520c3197d87003)

**Comment by sanatmpa1** [comment]
@Space-Fighter,

I passed a random image and I am getting a different error when running [this code](https://gist.github.com/Space-Fighter/7f926edf96260758f901b06ae5a72157), `ValueError: Input tensors to a Functional must come from 'tf.keras.Input'` . Please find the [gist here](https://colab.sandbox.google.com/gist/sanatmpa1/0869b83260679fe1e824b1ebfbc032b6/github_52184.ipynb).

Instead of sharing the code in a seperate link, Can you share a colab gist which can expedite the trouble shooting process?

**Comment by Space-Fighter** [comment]
> > @Space-Fighter
> > As per the [script](https://pastebin.com/kCr8DDLk) that you provided, You have provided input `x` as `None` in this line `c1, p1 = down_block(None, f[0], input = True) #128 -> 64` and so while trying to reproduce your code, I am getting the error `TypeError: Inputs to a layer should be tensors. Got: None` [gist here](https://colab.research.google.com/gist/sanatmpa1/e92bbc22183fc6d32e8fa8344ef0bb4e/52184.ipynb). Can you provide a github gist to reproduce the issue reported?
> 
> @sanatmpa1 Minor correction in the post i mentioned the following line - ![image](https://user-images.githubusercontent.com/63576756/136048511-c37f4e93-a6a9-41b0-aced-7834b057de79.png) but i mistakenly sent my old script - ![image](https://user-images.githubusercontent.com/63576756/136048696-6c9ebf1d-17e2-417c-afdf-54323d328824.png)
> 
> Here is the script - [SCRIPT](https://gist.github.com/Space-Fighter/53d7a25dd50278c9106103dc6bb5afbd)
> 
> I have updated this in the post too
> 
> Regarding the following error - `TypeError: Inputs to a layer should be tensors. Got: None`
> 
> After doing the correction as mentioned above in the google colab you sent me I get a new error -
> 
> ![image](https://user-images.githubusercontent.com/63576756/136051306-3e2979a7-4939-45e1-9ceb-7c2efe95730e.png)
> 
> But in my local machine i get the same error as i have specified in my post -
> 
> ![image](https://user-images.githubusercontent.com/63576756/136051842-25c9dbe0-7f2e-4f02-a50e-06da9489e833.png)

@sanatmpa1 As i see in the post quoted above i seem to get a different error in my local machine and a diffrent one in colab. 

As i see the version of tensorflow in colab is - 
![image](https://user-images.githubusercontent.com/63576756/139585230-ba01ba09-85b5-48b9-bb23-74b6424ad44b.png)

and in my machine it is - 

![image](https://user-images.githubusercontent.com/63576756/139585385-66cdae7b-431a-44e2-951c-bc51ac2fcbd4.png)

**Comment by sanatmpa1** [comment]
@Space-Fighter,

Can you try updating TF to latest stable version i.e 2.6.0 in your local machine as well? as 2.1.0 is an older version and there are many different problems fixed and features added in `2.6.0`
